pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,139 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsReady is the heartbeat endpoint for gRPC.
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
+}
+
+/**
+ * EventTime represents the event time of the datum.
+ */
+message EventTime {
+  google.protobuf.Timestamp event_time = 1;
+}
+
+/**
+ * ReadRequest is the request for reading datum stream from user defined source.
+ */
+message ReadRequest {
+  message Request {
+    // num_records is the number of records to read.
+    uint64 num_records = 1;
+  }
+  Request request = 1;
+}
+
+/**
+ * ReadResponse represents a list of datum response elements.
+ */
+message ReadResponse {
+  message Result {
+    // elements is a list of datum response elements.
+    repeated DatumResponse elements = 1;
+  }
+  Result result = 1;
+}
+
+/**
+ * DatumResponse represents a datum response element.
+ */
+message DatumResponse {
+  message Result {
+    // payload is the payload of the datum.
+    bytes payload = 1;
+    // offset is the offset information of the datum.
+    Offset offset = 2;
+    // keys is an optional list of keys associated with the datum.
+    // Key is the "key" attribute in (key,value) as in the map-reduce paradigm.
+    // We add this optional field to support the use case where the user defined source can provide keys for the datum.
+    // e.g. Kafka and Redis Stream message usually include information about the keys.
+    repeated string keys = 3;
+    // event_time is the time associated with each datum.
+    // We add this optional field to support the use case where the user defined source can provide event time for the datum.
+    // e.g. Kafka and Redis Stream message usually include information about the event time.
+    EventTime event_time = 4;
+  }
+  Result result = 1;
+}
+
+/**
+ * AckRequest is the request for acknowledging datum.
+ * It takes a list of offsets to be acknowledged.
+ */
+message AckRequest {
```

Comment:
till now, we prefix with `xxxList` if it is a repeated item, i am not opinionated on either, but we should agree and follow the same conventions.

Reply:
I have strong opinions on not using `xxxList` because it leaves users the impression that the message only contains one single `repeated` field. In the future, we might want to add more fields to enrich the message. e.g. (might not completely making sense) for `AckRequest`, if we want to ack a list of offsets that belong to one of the partitions from the source, we might want to add to the Request message another field `string partition_id = 2`. For that case, I think `Request` is a better name than `RequestList`.